include/rocksdb/db.h

```diff
@@ -156,7 +159,7 @@ class DB {
   //
   // Caller must delete *dbptr when it is no longer needed.
   static Status Open(const Options& options, const std::string& name,
-                     DB** dbptr);
+                     UniquePtrOut<DB> dbptr);
```

Comment:
I'm all for making the ownership semantics clear in the public API. However, I'm a bit concerned about introducing a custom smart pointer-like type, since there are a lot of ways it could be accidentally misused by application code. (I realize `UniquePtrOut` is only meant to be implicitly constructed when `Open` is called with a pointer to either a raw or a smart pointer but there's no way we can guarantee that.)

An alternative solution to consider: introduce new overloads that take `unique_ptr`s and turn the current methods into thin wrappers around the `unique_ptr` versions. (We could allocate a `unique_ptr` on the stack in the wrapper, call the smart pointer version, and then call `release` on the `unique_ptr` and copy the raw pointer value to the out parameter.) In addition, we could mark the raw pointer versions deprecated and remove them in 7.0.

Reply:
I was hoping to avoid "overload bloat." Perhaps it is the lesser of evils.